File: src/collector/first_run.ts
import { stat } from 'node:fs/promises';
import { join } from 'node:path';
import { existsSync } from 'node:fs';
import type { Adapter } from '../adapters/adapter.js';
import type { Repository } from '../store/repository.js';
import type { PricingTable } from '../pricing/types.js';
import { ingestFile } from './pipeline.js';
import { ingestHistoryFile } from '../adapters/claude_code/history_jsonl.js';
import { ClaudeCodeAdapter } from '../adapters/claude_code/index.js';

export interface FirstRunOpts { recentDays: number; }

export interface IngestStatus { foregroundComplete: boolean; pending: number; processed: number; total: number; }

export function runFirstPassIngest(adapters: Adapter[], repo: Repository, table: PricingTable, opts: FirstRunOpts) {
  const cutoff = Date.now() - opts.recentDays * 86_400_000;
  let processed = 0, total = 0;
  let foregroundComplete = false;

  const foregroundDone = (async () => {
    const recent: { adapter: Adapter; file: string }[] = [];
    const older: { adapter: Adapter; file: string }[] = [];
    for (const a of adapters) {
      for (const f of await a.discoverSessionFiles()) {
        try {
          const st = await stat(f);
          if (st.mtimeMs >= cutoff) recent.push({ adapter: a, file: f });
          else older.push({ adapter: a, file: f });
        } catch { /* skip */ }
      }
    }
    total = recent.length + older.length;
    for (const { adapter, file } of recent) {
      try { await ingestFile(adapter, file, repo, table); }
      catch (e) {
        repo.recordParseError({ file, byte_offset: -1, reason: `[ingest] ${e instanceof Error ? e.message : String(e)}`, raw_line_truncated: '' });
      }
      processed++;
    }

    // Slice 1: ingest ~/.claude/history.jsonl for the Prompts page during
    // the foreground phase. It's a small single file (~1500 lines for a
    // heavy user) so reading it eagerly doesn't push the foreground time
    // past acceptable. Done after recent sessions so the prompts have
    // session candidates to link against.
    for (const a of adapters) {
      if (a.agent === 'claude_code') {
        const historyPath = join(a.rootPath(), 'history.jsonl');
        if (existsSync(historyPath)) {
          try { await ingestHistoryFile(historyPath, repo); }
          catch (e) {
            repo.recordParseError({
              file: historyPath, byte_offset: -1,
              reason: `[history] ${e instanceof Error ? e.message : String(e)}`,
              raw_line_truncated: '',
            });
          }
        }
      }
    }

    foregroundComplete = true;
    return { older };
  })();

  const backfillDone = (async () => {
    const { older } = await foregroundDone;
    for (const { adapter, file } of older) {
      try { await ingestFile(adapter, file, repo, table); }
      catch (e) {
        repo.recordParseError({ file, byte_offset: -1, reason: `[ingest] ${e instanceof Error ? e.message : String(e)}`, raw_line_truncated: '' });
      }
      processed++;
    }

    // Slice 1+2: backfill tool_calls and transcript segments for sessions
    // ingested before those slices shipped (or before either was added to
    // a given session's data). We re-ingest those files from offset 0 —
    // the upsert pattern means turns and offsets stay consistent, but
    // tool_calls and segments populate where they were missing. Cap at
    // 200 sessions per startup so a huge backlog doesn't dominate the
    // background phase. The two backfills share a single re-ingest so we
    // never re-walk the same file twice.
    await backfillMissingDerivedData(adapters, repo, table);
  })();

  return { foregroundDone, backfillDone, status: (): IngestStatus => ({ foregroundComplete, pending: total - processed, processed, total }) };
}

// Re-ingest sessions that don't yet have any derived data (tool_calls
// rows or transcript_segments rows). We reset their byte offset to 0
// and let the normal pipeline re-walk; existing turns upsert in place
// (no double-count) and tool_calls/segments populate as a side effect.
// Top-level sessions only — sub-agent files are walked as part of their
// parent's re-ingest. Cap at 200 candidates total so this never
// dominates the background phase on a fresh upgrade.
async function backfillMissingDerivedData(adapters: Adapter[], repo: Repository, table: PricingTable): Promise<void> {
  const missingTools = repo.sessionsMissingToolCalls(200);
  const missingSegs = repo.sessionsMissingSegments(200);
  const ids = new Set<string>([...missingTools.map(c => c.id), ...missingSegs.map(c => c.id)]);
  const candidates = [...ids].slice(0, 200).map(id => ({ id }));
  if (candidates.length === 0) return;

  // Map session id → likely file path. We can derive it from
  // discoverSessionFiles for each adapter and match by basename, since the
  // session id is `${agent}:${basename(file, '.jsonl')}`.
  const fileByBasename = new Map<string, { adapter: Adapter; file: string }>();
  for (const a of adapters) {
    if (a.agent !== 'claude_code') continue;
    // Reuse discoverSessionFiles. ClaudeCodeAdapter only returns top-level
    // session files, which is what we want here.
    if (!(a instanceof ClaudeCodeAdapter)) continue;
    for (const f of await a.discoverSessionFiles()) {
      const base = f.split(/[\\/]/).pop()?.replace(/\.jsonl$/, '');
      if (base) fileByBasename.set(base, { adapter: a, file: f });
    }
  }

  for (const { id } of candidates) {
    // Session ids are formatted "claude_code:<basename>" for top-level
    // sessions. Ignore anything with a slash (sub-agent rollup) — those
    // get backfilled when their parent re-ingests.
    if (id.includes('/')) continue;
    const colon = id.indexOf(':');
    if (colon < 0) continue;
    const native = id.slice(colon + 1);
    const match = fileByBasename.get(native);
    if (!match) continue;
    // Reset offset so we re-walk the whole file, then ingest.
    repo.setFileOffset(match.file, 0);
    try { await ingestFile(match.adapter, match.file, repo, table); }
    catch (e) {
      repo.recordParseError({
        file: match.file, byte_offset: -1,
        reason: `[backfill-tools] ${e instanceof Error ? e.message : String(e)}`,
        raw_line_truncated: '',
      });
    }
  }
}

File: python/argus/collector/first_run.py
"""First-pass ingest: walk every adapter's files, recent first.

Recent files run synchronously in the foreground so the dashboard is
useful immediately. Older files run in a background ThreadPoolExecutor.
"""
from __future__ import annotations

import logging
import threading
import time
from concurrent.futures import Future, ThreadPoolExecutor
from dataclasses import dataclass
from pathlib import Path

from ..adapters.base import Adapter
from ..pricing.types import PricingTable
from ..store.repository import Repository
from .pipeline import ingest_file

logger = logging.getLogger(__name__)


@dataclass
class IngestStatus:
    foreground_complete: bool
    pending: int
    processed: int
    total: int


class FirstRunHandle:
    """Returned by ``run_first_pass_ingest``; exposes foreground/backfill futures."""

    def __init__(self) -> None:
        self._processed = 0
        self._total = 0
        self._foreground_complete = False
        self._lock = threading.Lock()
        self._foreground_done = threading.Event()
        self._backfill_done = threading.Event()

    def _inc(self) -> None:
        with self._lock:
            self._processed += 1

    def status(self) -> IngestStatus:
        with self._lock:
            return IngestStatus(
                foreground_complete=self._foreground_complete,
                pending=max(0, self._total - self._processed),
                processed=self._processed,
                total=self._total,
            )

    def wait_foreground(self, timeout: float | None = None) -> bool:
        return self._foreground_done.wait(timeout)

    def wait_backfill(self, timeout: float | None = None) -> bool:
        return self._backfill_done.wait(timeout)


def run_first_pass_ingest(
    adapters: list[Adapter],
    repo: Repository,
    table: PricingTable,
    *,
    recent_days: int = 30,
) -> FirstRunHandle:
    """Kick off ingest. Recent files run inline; older files in a thread.

    Returns immediately with a handle whose ``status()`` is pollable, and
    whose ``wait_foreground()`` / ``wait_backfill()`` block until each
    phase finishes.
    """
    cutoff = time.time() - recent_days * 86_400
    handle = FirstRunHandle()

    # Phase 1 (foreground, sync, in the calling thread).
    recent: list[tuple[Adapter, Path]] = []
    older: list[tuple[Adapter, Path]] = []
    for a in adapters:
        for f in a.discover_session_files():
            try:
                mtime = f.stat().st_mtime
            except OSError:
                continue
            (recent if mtime >= cutoff else older).append((a, f))

    with handle._lock:
        handle._total = len(recent) + len(older)

    for adapter, file in recent:
        try:
            ingest_file(adapter, file, repo, table)
        except Exception as e:  # noqa: BLE001
            repo.record_parse_error(
                {
                    "file": str(file),
                    "byte_offset": -1,
                    "reason": f"[ingest] {e}",
                    "raw_line_truncated": "",
                }
            )
        handle._inc()

    # Also ingest adapter-specific extras (e.g., history.jsonl) during the
    # foreground phase so they're available to the dashboard immediately.
    for a in adapters:
        for extra in a.extra_watch_paths():
            try:
                a.ingest_extra(extra, repo)
            except Exception as e:  # noqa: BLE001
                repo.record_parse_error(
                    {
                        "file": str(extra),
                        "byte_offset": -1,
                        "reason": f"[history] {e}",
                        "raw_line_truncated": "",
                    }
                )

    with handle._lock:
        handle._foreground_complete = True
    handle._foreground_done.set()

    # Phase 2 (background) — older files + missing-data backfill.
    def _background() -> None:
        for adapter, file in older:
            try:
                ingest_file(adapter, file, repo, table)
            except Exception as e:  # noqa: BLE001
                repo.record_parse_error(
                    {
                        "file": str(file),
                        "byte_offset": -1,
                        "reason": f"[ingest] {e}",
                        "raw_line_truncated": "",
                    }
                )
            handle._inc()
        _backfill_missing_derived_data(adapters, repo, table)
        handle._backfill_done.set()

    threading.Thread(target=_background, name="argus-firstrun-bg", daemon=True).start()
    return handle


def _backfill_missing_derived_data(
    adapters: list[Adapter], repo: Repository, table: PricingTable
) -> None:
    """Re-ingest sessions missing tool_calls / segments after a slice upgrade."""
    missing_tools = repo.sessions_missing_tool_calls(200)
    ids: set[str] = {c["id"] for c in missing_tools}
    if repo.is_search_indexing_enabled():
        for c in repo.sessions_missing_segments(200):
            ids.add(c["id"])
    candidates = sorted(ids)[:200]
    if not candidates:
        return

    # session_id "claude_code:<basename>" → file path lookup.
    file_by_basename: dict[str, tuple[Adapter, Path]] = {}
    for a in adapters:
        for f in a.discover_session_files():
            file_by_basename[f.stem] = (a, f)

    for id_ in candidates:
        if "/" in id_:  # sub-agent rollup ids — walked via parents
            continue
        colon = id_.find(":")
        if colon < 0:
            continue
        native = id_[colon + 1 :]
        match = file_by_basename.get(native)
        if match is None:
            continue
        adapter, file = match
        repo.set_file_offset(str(file), 0)
        try:
            ingest_file(adapter, file, repo, table)
        except Exception as e:  # noqa: BLE001
            repo.record_parse_error(
                {
                    "file": str(file),
                    "byte_offset": -1,
                    "reason": f"[backfill-tools] {e}",
                    "raw_line_truncated": "",
                }
            )
